cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/response/RightsizeResult.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2021 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+ */
+
+package com.linkedin.kafka.cruisecontrol.servlet.response;
+
+import com.google.gson.Gson;
+import com.linkedin.cruisecontrol.servlet.parameters.CruiseControlParameters;
+import com.linkedin.kafka.cruisecontrol.config.KafkaCruiseControlConfig;
+import com.linkedin.kafka.cruisecontrol.detector.ProvisionerState;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.linkedin.kafka.cruisecontrol.servlet.response.ResponseUtils.*;
+
+
+@JsonResponseClass
+public class RightsizeResult extends AbstractCruiseControlResponse {
+  @JsonResponseField
+  protected static final String NUM_BROKERS_TO_ADD = "numBrokersToAdd";
+  @JsonResponseField
+  protected static final String PARTITION_COUNT = "partitionCount";
+  @JsonResponseField
+  protected static final String TOPIC = "topic";
+  @JsonResponseField
+  protected static final String PROVISIONER_STATE = "provisionerState";
+  @JsonResponseField
+  protected static final String PROVISIONER_SUMMARY = "provisionerSummary";
+
+  protected final int _numBrokersToAdd;
+  protected final int _partitionCount;
+  protected String _topic;
+  protected String _provisionState;
+  protected String _provisionSummary;
+
+  public RightsizeResult(int numBrokersToAdd,
+                         int partitionCount,
+                         String topic,
+                         ProvisionerState provisionerState,
+                         KafkaCruiseControlConfig config) {
+    super(config);
+
+    _numBrokersToAdd = numBrokersToAdd;
+    _partitionCount = partitionCount;
+    _topic = topic;
+    _provisionState = provisionerState.state().toString();
+    _provisionSummary = provisionerState.summary();
+  }
+
+  protected String getJSONString() {
+    Map<String, Object> jsonStructure = new HashMap<>(5);
+    jsonStructure.put(NUM_BROKERS_TO_ADD, _numBrokersToAdd);
+    jsonStructure.put(PARTITION_COUNT, _partitionCount);
+    if (_topic != null && !_topic.isEmpty()) {
+      jsonStructure.put(TOPIC, _topic);
+    }
+    if (_provisionState != null && !_provisionState.isEmpty()) {
+      jsonStructure.put(PROVISIONER_STATE, _provisionState);
+    }
+    if (_provisionSummary != null && !_provisionSummary.isEmpty()) {
+      jsonStructure.put(PROVISIONER_SUMMARY, _provisionSummary);
+    }
```

Comment:
I am a little confused about what is expected to be `optional` and what is supposed to be `required`.
Based on the sanity checks in `ProvisionRecommendation`, a provision recommendation can be built only in certain ways.

For example, `ProvisionRecommendation#sanityCheckResources` ensures the following:
```
   * <ul>
   *   <li>exactly one resource type is set</li>
   *   <li>if the resource type is partition, then the cluster is under provisioned</li>
   *   <li>if the resource type is partition, then the corresponding topic must be specified; otherwise, the topic cannot be specified</li>
   * </ul>
```
So `_numBrokersToAdd` and `_partitionCount` are guaranteed to be _not_ specified together. So in a rightsize request, user can either specify num brokers or num partitions but not both.

Similarly, if the user specifies `_partitionCount`, then s/he must specify the topic.
Also, is there any case that `_provisionState` is expected to be null or empty (I don't think it is possible)?
Can _provisionSummary ever be `null`? Would it be empty? If it can be empty, is there a special meaning to it that makes us omit the response?

Also `ProvisionerState` has a `toString()` method. Can we use that in response to `PROVISIONER_STATE` and drop seperate `_provisionState` and `_provisionSummary` responses here?

Reply:
We've added the case that if the rightsize method is called with an empty provision recommendation, it will return `null` to reduce the amount of logging. 
For this endpoint however, I'm adding a sanity check to throw `IllegalArgumentException` if neither the broker nor partition are specified. Therefore, the `RightsizeResults#provisionerState` should never be `null` since we reject empty recommendations